quickwit-cluster/docs/kubernetes-discovery.md

```diff
@@ -0,0 +1,195 @@
+garbage collection, to support very high churn scenarios.
+
+## Illustrative Quickwit example release topology
+
+In [Helm terminology](https://helm.sh/docs/glossary/), a chart defines templates that the user will
+configure and render into a release (a set of k8s entity manifests) and apply into one k8s cluster.
+Applying several releases of the chart with different options, users can deploy several decoupled Quickwit clusters.
+Resource names are usually prefixed with the configurable release name to avoid conflicts.
+
+Defining the topology of the official Quickwit chart is out of scope for this document (especially choosing between
+deployments and statefulsets), but we will consider this potential topology (inspired from Bitnami's
+[Cassandra chart](https://github.com/bitnami/charts/tree/master/bitnami/cassandra/templates) and my experience)
+as an illustration for discussions on the proposed solution:
+
+- a `$release-indexing` deployment for an elastic pool of indexer nodes, that can be quickly auto-scaled based on volume
+- a `$release-indexing` service, to provide a DNS endpoint for the push API. This service uses a label selector to 
+  expose pods of the `$release-indexing` deployment once [they are ready](https://blog.colinbreck.com/kubernetes-liveness-and-readiness-probes-how-to-avoid-shooting-yourself-in-the-foot/)
+- a `$release-search` statefulset, for a pool of searcher nodes. Statefulset would allow prioritizing disk caching
+  (with persistent disk volumes) over elasticity (they are slower to up/down scale), if the read volume is steady
+- a `$release-search` service, to provide a DNS endpoint for the search API. It uses a label selector to
+  expose ready pods of the `$release-search` statefulset
+- a `$release-headless` service, to be used for discovery (see next section). It will be set:
+  - as [headless](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services), to allow
+    enumerating all pods at the DNS level
+  - with `publishNotReadyAddresses: true` to expose all pods, even before they are ready
+  - with a label selector matching both the `$release-indexing` and `$release-search` pods, as a unified cluster
+
+## Proposed solution
+
+The following action items are ordered by priority, and could be independently implemented in several releases.
+Alternatives are discussed in the next section, with a pros/cons comparison.
+
+### 1a. Source the sibling pods as seeds via DNS enumeration on startup
+
+A new `peer_seed_service` configuration option is added to Quickwit. It is exclusive to the existing `peer_seeds`,
+and setting both will result on a validation error. `quickwit-cluster::start_cluster_service` is updated to check
+this option. If it is not empty:
+
+- a DNS enumeration request is issued, that will return the IP addresses of all the live pods in the cluster,
+- if the DNS request fails or if the result is empty, Quickwit fails to start. Kubernetes will restart the
```

Comment:
interesting. Is it common for the result to be empty? I think our own IP address will be there usually but I guess this is not guaranteed.

Reply:
When the chart is first applied and the first pods come up (which can be fast if capacity is already available in the cluster), they could run the request before the DNS entry is created and propagated to the servers.

In that case, failing and retrying later will allow the cluster to eventually bootstrap without human intervention. We could decide to implement a retry logic, but we would be adding complexity for the same outcome. I prefer to rely on k8s's crash loop backoff mechanism, as it is more visible to cluster operators.